### PR TITLE
Streamline expression context generation

### PR DESCRIPTION
**Pull request by m-kuhn**
Whenever an object is able to generate an expression context it
implements the method createExpressionContext() declared in
QgsExpressionContextGenerator.

This makes a cleaner API and allows using QgsFieldExpressionWidget and
QgsDataDefinedButton from python because standard OO programming
approaches are used instead of callbacks and void pointers.

CC @nyalldawson aka mr. expression context

PS: misses api break docs

### CODE REVIEW

File: python/core/qgsexpressioncontextgenerator.sip [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+/**
+ * \ingroup core
+ * Abstract interface for generatign an expression context.
```
Comment (nyalldawson) [inline python/core/qgsexpressioncontextgenerator.sip]: generatign->generating

File: python/core/qgsexpressioncontextgenerator.sip [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * You need to implement this interface in a class and register this class with
+ * QgsFieldExpressionWidget::registerExpressionGenerator().
```
Comment (nyalldawson) [inline python/core/qgsexpressioncontextgenerator.sip]: should the data defined button be explicitly mentioned here too?

Comment (m-kuhn) [inline python/core/qgsexpressioncontextgenerator.sip]: I can do that

**Comment by nyalldawson** [comment]
Sorry - I didn't get a chance to finish going through this. Tomorrow :)

**Comment by m-kuhn** [comment]
Ok, I'll wait for your feedback.
I'll fix up the tests (working on colored output for doc and sip tests now) and the two things you mentioned today.

**Comment by m-kuhn** [comment]
@nyalldawson do you have any idea why the sip test fails? Bindings are there and I cannot reproduce this locally.

**Pull request merged**